File: Sources/Data/FileSubscriptionRepository.swift
import Foundation
import Domain

public actor FileSubscriptionRepository: SubscriptionRepository {
    private let store: LocalJSONStore

    public init() {
        self.store = LocalJSONStore()
    }

    public func fetchAll() async throws -> [Subscription] {
        try await store.load()
            .filter(\.isActive)
            .sorted(by: { $0.nextBillingDate < $1.nextBillingDate })
    }

    public func create(_ subscription: Subscription) async throws {
        try subscription.validate()
        var current = try await store.load()
        current.append(subscription)
        try await store.save(current)
    }

    public func update(_ subscription: Subscription) async throws {
        try subscription.validate()
        var current = try await store.load()
        guard let index = current.firstIndex(where: { $0.id == subscription.id }) else {
            throw DomainError.notFound
        }
        current[index] = subscription
        try await store.save(current)
    }

    public func delete(id: UUID) async throws {
        var current = try await store.load()
        guard current.contains(where: { $0.id == id }) else {
            throw DomainError.notFound
        }
        current.removeAll(where: { $0.id == id })
        try await store.save(current)
    }

    public func upcoming(within days: Int, from now: Date = .now) async throws -> [Subscription] {
        let future = Calendar.current.date(byAdding: .day, value: max(0, days), to: now) ?? now
        return try await fetchAll().filter { subscription in
            (subscription.nextBillingDate >= now) && (subscription.nextBillingDate <= future)
        }
    }
}

File: android/core/data/src/main/java/kr/co/cdd/payboard/core/data/repository/FileSubscriptionRepository.kt
package kr.co.cdd.payboard.core.data.repository

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext
import kr.co.cdd.payboard.core.data.local.LocalJsonStore
import kr.co.cdd.payboard.core.domain.model.Subscription
import kr.co.cdd.payboard.core.domain.repository.SubscriptionRepository
import java.time.Instant
import java.util.UUID

class FileSubscriptionRepository(
    context: Context,
) : SubscriptionRepository {
    private val store = LocalJsonStore(context.applicationContext)
    private val allSubscriptions = MutableStateFlow(store.load())

    override fun observeActive(): Flow<List<Subscription>> =
        allSubscriptions.map { subscriptions ->
            subscriptions
                .filter(Subscription::isActive)
                .sortedBy(Subscription::nextBillingDate)
        }

    override fun observeArchived(): Flow<List<Subscription>> =
        allSubscriptions.map { subscriptions ->
            subscriptions
                .filterNot(Subscription::isActive)
                .sortedByDescending(Subscription::updatedAt)
        }

    override suspend fun fetchAll(): List<Subscription> = withContext(Dispatchers.IO) {
        allSubscriptions.value.toList()
    }

    override suspend fun replaceAll(subscriptions: List<Subscription>) {
        withContext(Dispatchers.IO) {
            subscriptions.forEach(Subscription::validate)
            store.save(subscriptions)
            allSubscriptions.value = subscriptions
        }
    }

    override suspend fun create(subscription: Subscription) {
        persist {
            subscription.validate()
            add(subscription)
        }
    }

    override suspend fun update(subscription: Subscription) {
        persist {
            subscription.validate()
            replaceAll { current ->
                if (current.id == subscription.id) subscription else current
            }
        }
    }

    override suspend fun delete(id: UUID) {
        persist {
            removeAll { it.id == id }
        }
    }

    override suspend fun archive(id: UUID) {
        persist {
            replaceAll { current ->
                if (current.id == id) current.copy(isActive = false, updatedAt = Instant.now()) else current
            }
        }
    }

    override suspend fun restore(id: UUID) {
        persist {
            replaceAll { current ->
                if (current.id == id) current.copy(isActive = true, updatedAt = Instant.now()) else current
            }
        }
    }

    private suspend fun persist(transform: MutableList<Subscription>.() -> Unit) {
        withContext(Dispatchers.IO) {
            val next = allSubscriptions.value.toMutableList().apply(transform)
            store.save(next)
            allSubscriptions.value = next
        }
    }
}
